### PR TITLE
add introductions for scheduler component config api migrations

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

xref: https://github.com/kubernetes/website/issues/30063
part of: https://github.com/kubernetes/enhancements/issues/2901

### CODE REVIEW

File: content/en/docs/reference/scheduling/config.md [hunk 1]
```diff
@@ -255,10 +226,30 @@ the same configuration parameters (if applicable). This is because the scheduler
 only has one pending pods queue.
 {{< /note >}}
 
+## Migration introductions between api versions
```
Comment (sftim) [inline content/en/docs/reference/scheduling/config.md:229]: What does “introductions” mean here? I can't work it out.

File: content/en/docs/reference/scheduling/config.md [hunk 2]
```diff
@@ -255,10 +226,30 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+
+### v1beta2 -> v1beta3
```
Comment (sftim) [inline content/en/docs/reference/scheduling/config.md:231]: I'd put:
```suggestion
### Scheduler configuration migration: v1beta2 → v1beta3 {#migration-v1beta2-v1beta3}
```

File: content/en/docs/reference/scheduling/config.md [hunk 3]
```diff
@@ -255,10 +226,30 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+* Some plugins' weight has been modified for user preferences:
+  * `InterPodAffinity` to 2
+  * `NodeAffinity` to 2
+  * `TaintToleration` to 3
+* Support for configuration file version `v1beta1` is removed.
+* Scheduler plugin `ServiceAffinity` is removed.
+* The legacy scheduler policy config is removed, the associated flags `policy-config-file`, `policy-configmap`, `policy-configmap-namespace` and `use-legacy-policy-config` are also removed.
+
+### v1beta1 -> v1beta2
```
Comment (sftim) [inline content/en/docs/reference/scheduling/config.md:240]:
```suggestion
### Scheduler configuration migration: v1beta1 → v1beta2 {#migration-v1beta1-v1beta2}
```

I would order this _before_ the migration from v1beta2 to v1beta3.

File: content/en/docs/reference/scheduling/config.md (hunk 1 above)
Comment (kerthcet) [reply content/en/docs/reference/scheduling/config.md:229]: I just mean some explanations about the changes between different api versions.

File: content/en/docs/reference/scheduling/config.md (hunk 3 above)
Comment (kerthcet) [reply content/en/docs/reference/scheduling/config.md]: I would use the tabs as advised.

File: content/en/docs/reference/scheduling/config.md (hunk 2 above)
Comment (kerthcet) [reply content/en/docs/reference/scheduling/config.md]: changed to tabs already.

File: content/en/docs/reference/scheduling/config.md [hunk 4]
```diff
@@ -255,10 +226,34 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+* Scheduler plugin `ServiceAffinity` is deprecated, it's only configurable through the scheduler policy API and v1beta1 component config API. It's recommended to use `InterPodAffinity` plugin instead.
+* Scheduler plugin `NodePreferAvoidPods` is deprecated, it's only configurable through the scheduler policy API and v1beta1 component config API. It's recommended to use node taints instead.
+* A plugin enabled in a v1beta2 configuration file takes precedence over the default configuration for that plugin. This simplifies enabling default plugins with custom configuration without the necessary to explicitly disable those default plugins.
+* New parameter `defaultingType` added for `PodTopologySpread` plugin allowing to use k8s defined or user provided default constraints.
+* Add pre-checks for scheduler healthz and metrics bind address.
+* The deprecated CLI flags `--hard-pod-affinity-symmetric-weight`, `--algorithm-provider` and `--scheduler-name` have been removed from kube-scheduler. Use `ComponentConfig` instead to configure those parameters.
```
Comment (alculquicondor) [inline content/en/docs/reference/scheduling/config.md:240]: Uhm.... this is not related to v1beta1->v1beta2 migration. No need to mention.

File: content/en/docs/reference/scheduling/config.md [hunk 5]
```diff
@@ -255,10 +251,58 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+* Scheduler plugin `NodePreferAvoidPods` is deprecated, using [node taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) instead.
+* A plugin enabled in a v1beta2 configuration file takes precedence over the default configuration for that plugin. This simplifies enabling default plugins with custom configuration without the necessary to explicitly disable those default plugins.
+* New parameter `defaultingType` added for `PodTopologySpread` plugin allowing to use k8s defined or user provided default constraints.
+* Add pre-checks for scheduler healthz and metrics bind address:
+```
+1. If the value is nil, default to 0.0.0.0 and default scheduler port
+2. If there is a value set, attempt to split it. If it's just a port (ie, ":1234"), default to 0.0.0.0 with that port
+3. If splitting the value fails, check if the value is even a valid IP. If so, use that with the default port.
+    Otherwise leave the address as-is, it will be caught during validation.
+```
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply content/en/docs/reference/scheduling/config.md]: PTAL again @sftim , I have updated the docs as you advised, thanks a lot.

File: content/en/docs/reference/scheduling/config.md [hunk 6]
```diff
@@ -255,10 +251,65 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+  For example, if you previously used the `NodeResourceLeastAllocated` plugin, you
+  would instead use `NodeResourcesFit` (enabled by default) and add a `pluginConfig`
+  with a `scoreStrategy` that is similar to:
+  ```yaml
+  apiVersion: kubescheduler.config.k8s.io/v1beta2
+  kind: KubeSchedulerConfiguration
+  profiles:
+  - pluginConfig:
+    - args:
+        apiVersion: kubescheduler.config.k8s.io/v1beta2
```
Comment (alculquicondor) [inline content/en/docs/reference/scheduling/config.md:271]: these 2 lines are optional. Better just remove them for easy of use.

File: content/en/docs/reference/scheduling/config.md [hunk 7]
```diff
@@ -255,10 +251,65 @@ the same configuration parameters (if applicable). This is because the scheduler
...
 
+## Scheduler configuration migrations
+
+{{< tabs name="tab_with_md" >}}
+{{% tab name="v1beta1 → v1beta2" %}}
+* With the v1beta2 configuration version, you can use a new score extension for the
+  `NodeResourcesFit` plugin.
+  The new extension combines the functionalities of the `NodeResourcesLeastAllocated`,
+  `NodeResourcesMostAllocated` and `RequestedToCapacityRatio` plugins.
+  For example, if you previously used the `NodeResourceLeastAllocated` plugin, you
```
Comment (alculquicondor) [inline content/en/docs/reference/scheduling/config.md:262]: let's use `NodeResourcesMostAllocated` instead, because Least is the default strategy.

File: content/en/docs/reference/scheduling/config.md [hunk 8]
```diff
@@ -255,10 +251,65 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+          resources:
+          - name: cpu
+            weight: 1
+          type: LeastAllocated
+      name: NodeResourcesFit
+  ```
+
+* The scheduler plugin `NodeLabel` is deprecated; instead, use the [`NodeAffinity`](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/) plugin(enabled by default).
+
+* The scheduler plugin `ServiceAffinity` is deprecated; instead, use the `InterPodAffinity` plugin(enabled by default). You should also add a `pluginConfig` that is similar to:
```
Comment (alculquicondor) [inline content/en/docs/reference/scheduling/config.md:283]: I don't think users **need** to add any specific configuration. It doesn't need to be part of the migration guide.

File: content/en/docs/reference/scheduling/config.md [hunk 9]
```diff
@@ -255,10 +251,65 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+        kind: InterPodAffinityArgs
+        hardPodAffinityWeight: 2
+      name: InterPodAffinity
+  ```
+
+* The scheduler plugin `NodePreferAvoidPods` is deprecated; instead, use [node taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) to achieve the same outcome.
+
+* A plugin enabled in a v1beta2 configuration file takes precedence over the default configuration for that plugin. This simplifies enabling default plugins with custom configuration without the necessary to explicitly disable those default plugins.
+
+* New parameter `defaultingType` added for `PodTopologySpread` plugin allowing to use k8s defined or user provided default constraints.
```
Comment (alculquicondor) [inline content/en/docs/reference/scheduling/config.md:300]: This is present in v1beta1 too, so no need to mention.

File: content/en/docs/reference/scheduling/config.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/reference/scheduling/config.md:271]: Is there any way that this `apiVersion` field could protect you from a future breaking change / any way we might wish that folks _had_ set this so we could emit warnings etc?

File: content/en/docs/reference/scheduling/config.md (hunk 8 above)
Comment (sftim) [inline content/en/docs/reference/scheduling/config.md:283]: nit: Also a space before the `(` please.

File: content/en/docs/reference/scheduling/config.md [hunk 10]
```diff
@@ -255,10 +251,47 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+    - args:
+        scoringStrategy:
+          resources:
+          - name: cpu
+            weight: 1
+          type: MostAllocated
+      name: NodeResourcesFit
+  ```
+
+* The scheduler plugin `NodeLabel` is deprecated; instead, use the [`NodeAffinity`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity) plugin (enabled by default) to achieve similar behavior.
```
Comment (kbhawkey) [inline content/en/docs/reference/scheduling/config.md:279]: 👋 @kerthcet . Would you have time to modify the links? For example:
/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity
/docs/concepts/scheduling-eviction/taint-and-toleration/

File: content/en/docs/reference/scheduling/config.md [hunk 11]
```diff
@@ -255,10 +251,47 @@ the same configuration parameters (if applicable). This is because the scheduler
...
+
+* The scheduler plugin `ServiceAffinity` is deprecated; instead, use the [`InterPodAffinity`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) plugin (enabled by default) to achieve similar behavior.
+
+* The scheduler plugin `NodePreferAvoidPods` is deprecated; instead, use [node taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) to achieve similar behavior.
+
+* A plugin enabled in a v1beta2 configuration file takes precedence over the default configuration for that plugin. This simplifies enabling default plugins with custom configuration without the necessary to explicitly disable those default plugins.
```
Comment (kbhawkey) [inline content/en/docs/reference/scheduling/config.md:285]: nit: Could you clarify the second sentence on line 285 or possibly remove that sentence?
This simplifies enabling default plugins with custom configurations without explicitly disabling those default plugins?

**Comment by kbhawkey** [comment]
Page preview:
https://deploy-preview-30151--kubernetes-io-main-staging.netlify.app/docs/reference/scheduling/config/

File: content/en/docs/reference/scheduling/config.md (hunk 10 above)
Comment (kerthcet) [reply content/en/docs/reference/scheduling/config.md]: done. :). thanks for pointing this out.

File: content/en/docs/reference/scheduling/config.md (hunk 11 above)
Comment (kerthcet) [reply content/en/docs/reference/scheduling/config.md]: Rmoved.

**Comment by kerthcet** [comment]
reopen for mis-operations and squashed all the commits.
